docs: re-arrange upgrade guide (#12504)
grafana/loki: docs/sources/setup/upgrade/_index.md

## Main / Unreleased

## 3.0.0

Loki 3.0 is a major version increase and comes with several breaking changes.

Here is the shortlist of things we think most people may encounter:

>   * Structured metadata is enabled by default and requires `tsdb` and `v13` schema or Loki won't start. See [Structured Metadata, Open Telemetry, Schemas and Indexes](#structured-metadata-open-telemetry-schemas-and-indexes)

Reviewer: ```suggestion
  * Structured metadata is enabled by default and requires `tsdb` and `v13` schema or Loki won't start. Refer to [Structured Metadata, Open Telemetry, Schemas and Indexes](#structured-metadata-open-telemetry-schemas-and-indexes).
```

>   * `shared_store` config is removed. See [Removed `shared_store` and `shared_store_key_prefix` from shipper configuration](#removed-sharedstore-and-sharedstorekeyprefix-from-shipper-configuration)

Reviewer: This should fix this error.
```suggestion
  * The `shared_store` config is removed. Refer to [Removed `shared_store` and `shared_store_key_prefix` from shipper configuration](#removed-shared_store-and-shared_store_key_prefix-from-shipper-configuration).
```

>   * Loki now enforces a max line size of 256KB by default (you can disable this or increase this but this is how we run Loki). See [Changes to default configure values](#changes-to-default-configuration-values-in-30)

Reviewer: ```suggestion
  * Loki now enforces a max line size of 256KB by default (you can disable this or increase this but this is how Grafana Labs runs Loki). Refer to [Changes to default configure values](#changes-to-default-configuration-values-in-30).
```

>   * Loki now enforces a max label limit of 15 labels per series, down from 30. Extra labels inflate the size of the index and reduce performance, you should almost never need more than 15 labels. See [Changes to default configure values](#changes-to-default-configuration-values-in-30)

Reviewer: ```suggestion
  * Loki now enforces a max label limit of 15 labels per series, down from 30. Extra labels inflate the size of the index and reduce performance, you should almost never need more than 15 labels. Refer to [Changes to default configure values](#changes-to-default-configuration-values-in-30).
```

>   * Be aware of a lot of metric name changes See [Distributor metric changes](#distributor-metric-changes), [Embedded cache metric changes](#embedded-cache-metric-changes), and [Metrics namespace](#metrics-namespace)

Reviewer: ```suggestion
  * There are many metric name changes. Refer to [Distributor metric changes](#distributor-metric-changes), [Embedded cache metric changes](#embedded-cache-metric-changes), and [Metrics namespace](#metrics-namespace).
```

> If you would like to see if your existing config will work with Loki 3.0:

Reviewer: ```suggestion
If you would like to see if your existing configuration will work with Loki 3.0:
```

>   * In an empty directory on your computer, put your config into a file named `loki-config.yaml`

Reviewer: ```suggestion
1. In an empty directory on your computer, copy you configuration into a file named `loki-config.yaml`.
```

>   * Run this command from that directory `docker run --rm -t -v "${PWD}":/config grafana/loki:3.0.0 -config.file=/config/loki-config.yaml -verify-config=true`

Reviewer: Moving the command into a code block so that the copy/paste functionality is available.  (Not sure if having thee ticks twice will work in the comment, so might have to do this outside of GitHub UI?)
```suggestion
1. Run this command from that directory:
  
```bash
  docker run --rm -t -v "${PWD}":/config grafana/loki:3.0.0 -config.file=/config/loki-config.yaml -verify-config=true
```

```

>   * Note: if you introduce a new schema_config entry it may cause additional validation errors

Reviewer: ```suggestion
{{< admonition type="note" >}}
If you introduce a new `schema_config` entry it may cause additional validation errors.
{{< /admonition >}}
```

>     * Tip: if you configure `path_prefix` in the `common` config section this can help save a lot of configuration. See the [Common Config Docs](https://grafana.com/docs/loki/latest/configure/#common)

Reviewer: ```suggestion
{{< admonition type="tip" >}}
If you configure `path_prefix` in the `common` config section this can help save a lot of configuration. Refer to [Common Config Docs](https://grafana.com/docs/loki/latest/configure/#common)
{{< /admonition >}}
```

> Note: the **Helm chart** us gone through some significant changes and has a separate upgrade guide: [Upgrading to Helm 6.x](https://grafana.com/docs/loki/latest/setup/upgrade/upgrade-to-6x/)

Reviewer: ```suggestion
{{< admonition type="note" >}}
The **Helm chart** has gone through significant changes and has a separate upgrade guide: [Upgrading to Helm 6.x](https://grafana.com/docs/loki/latest/setup/upgrade/upgrade-to-6x/).
{{< /admonition >}}
```

## Main / Unreleased

## 3.0.0

{{% admonition type="note" %}}

> If you have questions about upgrading to Loki 3.0, please join us at https://slack.grafana.com/ in the channel `#loki-3`

Reviewer: ```suggestion
If you have questions about upgrading to Loki 3.0, please join us on the [community Slack(https://slack.grafana.com/) in the `#loki-3` channel.
```

> Or leave a comment on this [Github Issue](https://github.com/grafana/loki/issues/12506)

Reviewer: ```suggestion
Or leave a comment on this [Github Issue](https://github.com/grafana/loki/issues/12506).
```

  * The `shared_store` config is removed. Refer to [Removed `shared_store` and `shared_store_key_prefix` from shipper configuration](#removed-shared_store-and-shared_store_key_prefix-from-shipper-configuration).
  * Loki now enforces a max line size of 256KB by default (you can disable this or increase this but this is how Grafana Labs runs Loki). Refer to [Changes to default configure values](#changes-to-default-configuration-values-in-30).
  * Loki now enforces a max label limit of 15 labels per series, down from 30. Extra labels inflate the size of the index and reduce performance, you should almost never need more than 15 labels. Refer to [Changes to default configure values](#changes-to-default-configuration-values-in-30).
  * There are many metric name changes. Refer to [Distributor metric changes](#distributor-metric-changes), [Embedded cache metric changes](#embedded-cache-metric-changes), and [Metrics namespace](#metrics-namespace).

If you would like to see if your existing configuration will work with Loki 3.0:
1. In an empty directory on your computer, copy you configuration into a file named `loki-config.yaml`.

>   * Run this command from that directory: 

Reviewer: I meant to make this a little numbered procedure.
```suggestion
1. Run this command from that directory: 
```

```bash
docker run --rm -t -v "${PWD}":/config grafana/loki:3.0.0 -config.file=/config/loki-config.yaml -verify-config=true`
```

> **Note:** if you introduce a new schema_config entry it may cause additional validation errors

Reviewer: ```suggestion
{{< admonition type="note" >}}
If you introduce a new schema_config entry it may cause additional validation errors.
{{< /admonition >}}
```

> **Tip:** if you configure `path_prefix` in the `common` config section this can help save a lot of configuration. See the [Common Config Docs](https://grafana.com/docs/loki/latest/configure/#common)

Reviewer: ```suggestion
{{< admonition type="tip" >}}
If you configure `path_prefix` in the `common` config section this can help save a lot of configuration. Refer to the [Common Config Docs](https://grafana.com/docs/loki/latest/configure/#common).
{{< /admonition >}}

```

> The **Helm chart** us gone through some significant changes and has a separate upgrade guide: [Upgrading to Helm 6.x](https://grafana.com/docs/loki/latest/setup/upgrade/upgrade-to-6x/)

Reviewer: ```suggestion
The **Helm chart** has gone through some significant changes and has a separate upgrade guide: [Upgrading to Helm 6.x](https://grafana.com/docs/loki/latest/setup/upgrade/upgrade-to-6x/).
```

additionally the `Background` configuration also lest you set `writeback_size_limit` which can be used to set a maximum amount of memory to use for writeback objects vs a count of objects.

#### Legacy ingester shutdown handler is removed

The already deprecated handler `/ingester/flush_shutdown` is removed in favor of `/ingester/shutdown?flush=true`.

#### Ingester configuration `max_transfer_retries` is removed.

The setting `max_transfer_retries` (`-ingester.max-transfer-retries`) is removed in favor of the Write Ahead log (WAL).
It was used to allow transferring chunks to new ingesters when the old ingester was shutting down during a rolling restart.
Alternatives to this setting are:
- **A. (Preferred)** Enable the WAL and rely on the new ingester to replay the WAL.
  - Optionally, you can enable `flush_on_shutdown` (`-ingester.flush-on-shutdown`) to flush to long-term storage on shutdowns.

> - **B.** Manually flush during shutdowns via [the ingester `/shutdown?flush=true` endpoint]({{< relref "../../reference/api#flush-in-memory-chunks-and-shut-down" >}}).

Reviewer: We can use full URLs now.
```suggestion
- **B.** Manually flush during shutdowns via [the ingester `/shutdown?flush=true` endpoint](https://grafana.com/docs/loki/latest/reference/api/#flush-in-memory-chunks-and-shut-down).
```
